std/typecons.d

```diff
@@ -6896,3 +6897,503 @@ unittest
+}
+
+unittest
+{
+    class C { bool destroyed = false; ~this() { destroyed = true; } }
+    auto c = new C;
+    auto optClass = some(c);
+    assert(!optClass.empty);
+    optClass = none;
+    assert(optClass.empty);
+    assert(!c.destroyed);
+}
+
+pure nothrow @safe @nogc unittest
+{
+    auto nullOpt = some(null);
+    Option!(int*) pOpt = nullOpt;
+    assert(!pOpt.empty);
+    nullOpt = none;
+    assert(nullOpt.empty);
+    pOpt = nullOpt;
+    assert(pOpt.empty);
+}
+
+pure nothrow @safe @nogc unittest
+{
+    struct NonCopyable
+    {
+        @disable this();
+        @disable this(this);
+    }
+
+    Option!NonCopyable defaultConstructed;
+    Option!NonCopyable opt = NonCopyable.init;
+    opt = some(NonCopyable.init);
+    opt = option(NonCopyable.init);
+}
+
+pure nothrow @safe @nogc unittest
+{
+    Option!(const int) opt = some!(const int)(42);
+    static assert(!__traits(compiles, opt = some(24)));
+    assert(!opt.empty);
+    assert(opt.front == 42);
+    opt = none;
+    assert(opt.empty);
+}
+
+pure /+ nothrow +/ @safe /+ @nogc +/ unittest
+{
+    int a = 42;
+    long b = 24;
+    auto opt = some(a).orElse(some(b));
+    static assert(is(typeof(opt) == Option!long));
+    assert(opt.front == 42);
+
+    auto pOpt = Option!(int*)(none).orElse(Option!(immutable int*)(none));
+    static assert(is(typeof(pOpt) == Option!(const(int)*)));
+    assert(pOpt.empty);
+}
+
+/// Constructor function for $(LREF Option).
+/// See_Also: $(LREF Option.this) for detailed behavior
+Option!T option(T)(T value)
+{
+    static if (isMutable!T)
+    {
+        import std.algorithm.mutation : move;
+        return Option!T(move(value));
+    }
+    else return Option!T(value);
+}
+
+///
+pure nothrow @safe unittest
+{
+    Object p = null;
+    assert(option(p).empty);
+
+    auto p2 = new Object;
+    assert(!option(p2).empty);
+}
+
+private struct NoneType {}
+
+/// Special value representing an absence of value when constructing $(LREF Option).
+immutable NoneType none = NoneType();
+
+///
+pure @safe unittest
+{
+    Option!int opt = none;
+    assert(opt.empty);
+    opt = some(42);
+    assert(opt.assumeNonEmpty == 42);
+}
+
+/// Construct an $(LREF Option) with the given value.
+/// The new instance is $(I always non-empty), also when value is null.
+/// See_Also: To determine emptiness based on null, use $(LREF option).
+Option!T some(T)(T value)
+{
+    static if (isMutable!T)
+    {
+        import std.algorithm.mutation : move;
+        return Option!T(move(value), true);
+    }
+    else return Option!T(value, true);
+}
+
+///
+pure nothrow @nogc unittest
+{
+    void* p = null;
+    assert(!some(p).empty);
+
+    void* p2 = &p;
+    assert(!some(p2).empty);
+
+    Option!(void*) opt = some(null);
+    assert(!opt.empty);
+}
+
+/**
+ * Return the `front` or `back` of range when it is non-empty, or $(LREF none) otherwise.
+ * See_Also: $(LREF Option)
+ */
+Option!(ElementType!Range) frontOption(Range)(Range range)
```

Comment:
Won't this be wrong in the case that Range is a string and its front returns `dchar`?